lib/spack/spack/database.py

```diff
@@ -937,6 +937,41 @@ def activated_extensions_for(self, extendee_spec, extensions_layout=None):
                 continue
             # TODO: conditional way to do this instead of catching exceptions
 
+    def get_by_hash_local(self, dag_hash, default=None):
```

Comment:
I think this should take an optional `installed` argument, like `Database._query()`. I think we can probably use these functions later to refactor `Database._query()` to be much faster for concrete specs.

Reply:
Fixed